src/interfaces/cloud.rs

```diff
@@ -0,0 +1,345 @@
+    upsert_entries_key: Option<[u8; 16]>,
+    insert_chains_key: Option<[u8; 16]>,
+}
+
+impl Display for Token {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut keys = self.findex_master_key.to_vec();
+
+        if let Some(fetch_entries_key) = self.fetch_entries_key {
+            keys.push(0);
+            keys.extend(fetch_entries_key);
+        }
+
+        if let Some(fetch_chains_key) = self.fetch_chains_key {
+            keys.push(1);
+            keys.extend(fetch_chains_key);
+        }
+
+        if let Some(upsert_entries_key) = self.upsert_entries_key {
+            keys.push(2);
+            keys.extend(upsert_entries_key);
+        }
+
+        if let Some(insert_chains_key) = self.insert_chains_key {
+            keys.push(3);
+            keys.extend(insert_chains_key);
+        }
+
+        write!(f, "{}{}", self.index_id, base64::encode(keys))
+    }
+}
+
+impl FromStr for Token {
+    type Err = FindexErr;
+
+    fn from_str(token: &str) -> Result<Self, Self::Err> {
+        let (index_id, tail) = token.split_at(5);
+        let mut bytes = base64::decode(tail)
+            .map_err(|_| {
+                FindexErr::Other(format!(
+                    "token {token} is not a valid base64 encoded string"
+                ))
+            })?
+            .into_iter();
+
+        let findex_master_key =
+            KeyingMaterial::try_from_bytes(&bytes.next_chunk::<16>().map_err(|_| {
+                FindexErr::Other(
+                    "the token is too short, cannot read the Findex master key".to_owned(),
+                )
+            })?)?;
+
+        let mut token = Token {
+            index_id: index_id.to_owned(),
+            findex_master_key,
+
+            fetch_entries_key: None,
+            fetch_chains_key: None,
+            upsert_entries_key: None,
+            insert_chains_key: None,
+        };
+
+        while let Some(prefix) = bytes.next() {
+            let key = Some(bytes.next_chunk::<16>().map_err(|_| {
+                FindexErr::Other(format!(
+                    "the token is too short, expecting 16 bytes after the prefix {prefix}"
+                ))
+            })?);
+
+            if prefix == 0 {
+                token.fetch_entries_key = key;
+            } else if prefix == 1 {
+                token.fetch_chains_key = key;
+            } else if prefix == 2 {
+                token.upsert_entries_key = key;
+            } else if prefix == 3 {
+                token.insert_chains_key = key;
+            } else {
+                return Err(FindexErr::Other(format!(
+                    "the token contains a unknown prefix {prefix}"
+                )));
+            }
+        }
+
+        Ok(token)
+    }
+}
+
+impl Token {
+    pub fn random_findex_master_key(
+        index_id: String,
+        fetch_entries_key: [u8; 16],
+        fetch_chains_key: [u8; 16],
+        upsert_entries_key: [u8; 16],
+        insert_chains_key: [u8; 16],
```

Comment:
Could you take optional parameters here?

I understand you prefer to create a full object in one client call, but the wrapping should be done at the interface level I think.

Reply:
I think there is no real sense to generate a new token with not all permissions. This operation is done before the usage of the index (since before this call there is no Findex Master key) and the user calling this function should have all the permissions on his index (since he is the one generating the Findex Master Key). Do you agree?